### PR TITLE
New faceted search results component

### PR DESCRIPTION
**Pull request by ke4**
This new component is going to replace the current `scxa-faceted-search-result` atlas component.
The current search component loaded all the results into memory and did the search result modification on that sometimes huge data structure. That was not ideal and sometimes we even got a timeout error when the time to return the response was longer than the allowed timeout value of our firewall.

In this change we are using separate endpoints for each of the filter components that was created in our backend previously:

- marker genes (show only marker genes) - checkbox (group)
- species (checkbox group)
- inferred cell type (drop down with multi selection)
- organism part (drop down with multi selection)

Any changes in the filters triggers to reload/repopulate the values of the other filters and the search result.
I also did a smaller optimisation for limiting the number of REST calls: the currently modified filter component is not going to call its endpoint to reload its value as it is not necessary, because we have them already.

On the backend we are about to start a ticket to make the search result paginated. When that is ready we should make sure that this frontend component could perfectly can communicate with that modified endpoint and can handle the search result, too.

Suggestion for improvement in following the PRs:
- adding a maximum 2 or 3 seconds of waiting time before executing/triggering the REST calls. That would make this component a bit more user-friendly and also it would limit the number of REST calls. I created a ticket in our backlog for this issue: [Adding waiting time before triggering the REST calls in the new faceted search result component](https://github.com/ebi-gene-expression-group/atlas-components/issues/139)

### CODE REVIEW

File: packages/scxa-faceted-search-results-v5/src/FilterSidebar.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import CheckboxFacetGroup from './facetgroups/CheckboxFacetGroup'
+import {withFetchLoader} from "@ebi-gene-expression-group/atlas-react-fetch-loader";
+import MultiselectDropdownFacetGroup from "./facetgroups/MultiselectDropdownFacetGroup";
+
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/src/FilterSidebar.js:6]: Quotes and semicolon style.

File: packages/scxa-faceted-search-results-v5/src/FilterSidebar.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    return {
+      checkboxFacetGroups: checkboxFacetGroups,
+      multiSelectDropdownFacetGroups: multiSelectDropdownFacetGroups
+    }
+  }
+
+  render() {
+    const { host, onChange } = this.props
+    const queryParamsObject = this.state.queryParams
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/src/FilterSidebar.js:53]: I’ve seen `queryParams` renamed to `queryStringObject`, `queryParamsObject` (here) or just `queryParams`. I would settle on a single name. What’s wrong with just `queryParams`?

File: packages/scxa-faceted-search-results-v5/src/FilterSidebar.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import CheckboxFacetGroup from './facetgroups/CheckboxFacetGroup'
+import {withFetchLoader} from "@ebi-gene-expression-group/atlas-react-fetch-loader";
+import MultiselectDropdownFacetGroup from "./facetgroups/MultiselectDropdownFacetGroup";
+
+const CheckboxFacetGroupWithFetchLoader = withFetchLoader(CheckboxFacetGroup)
+const MultiselectDropdownFacetGroupWithFetchLoader = withFetchLoader(MultiselectDropdownFacetGroup)
+
+class FilterSidebar extends React.Component {
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/src/FilterSidebar.js:10]: It doesn’t seem that `FilterSidebar` is stateful. We should get rid of the state and rewrite it as a function to make that explicit.

File: packages/scxa-faceted-search-results-v5/src/FilterList.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import PropTypes from 'prop-types'
+import _ from 'lodash'
+class FilterList extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      ascending: false,
+      sortTitle: props.sortTitle
+    }
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/src/FilterList.js:11]: I think that for the time being we should remove sorting. Having pagination makes it really unintuitive. As a side benefit, we can have a stateless `FilterList`.

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
This looks quite good! Please have a look at the detailed comments, there are several JavaScript and React stuff that should be improved or fixed.

File: packages/scxa-faceted-search-results-v5/cypress/component/CheckboxFacetGroup.cy.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    cy.mount(<CheckboxFacetGroup {...props }/>)
+    cy.get(`div.padding-bottom-xlarge h4 sup span`)
+      .should(`have.class`, `icon icon-generic`)
+      .should(`have.attr`, `data-tip`).then((dataTip) => {
+        expect(dataTip.toString()
+          .replace(`<span>`, ``)
+          .replace(`</span>`, ``)
+          .replace(`<br>`, ` `))
+          .to.equal(facetTooltip)
```
[the remark this answers is not in the document]
Comment (ke4) [reply packages/scxa-faceted-search-results-v5/cypress/component/CheckboxFacetGroup.cy.js:24]: It is the same, I was thinking to use a variable for it with a meaningful name, but it is OK for me also to not use a variable for it, as it is a simple one.

File: packages/scxa-faceted-search-results-v5/src/FacetedSearchContainer.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import FilterSidebar from './FilterSidebar'
+
+import CheckboxFacetGroupsDefaultProps from './facetgroups/CheckboxFacetGroupsDefaultProps'
+import MultiSelectDropdownFacetGroupsDefaultProps from "./facetgroups/MultiSelectDropdownFacetGroupsDefaultProps";
+import {withFetchLoader} from "@ebi-gene-expression-group/atlas-react-fetch-loader";
+import FilterList from "./FilterList";
```
[the remark this answers is not in the document]
Comment (ke4) [reply packages/scxa-faceted-search-results-v5/src/FacetedSearchContainer.js:9]: Ok all of these came from WebStorm automatic import (double quotes and semicolon). I am going to fix my settings and fix the code. Thanks for the reminder ;-)

File: packages/scxa-faceted-search-results-v5/src/facetgroups/CheckboxFacetGroup.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  label: PropTypes.string.isRequired,
+  disabled: PropTypes.bool.isRequired,
+  checked: PropTypes.bool.isRequired,
+  onChange: PropTypes.func.isRequired
+}
+
+class CheckboxFacetGroup extends React.Component {
+
+  getSelectedFacets(checkedFacets, changedFacetLabel) {
+    if (checkedFacets.indexOf(`true`) !== -1) {
```
[the remark this answers is not in the document]
Comment (ke4) [reply packages/scxa-faceted-search-results-v5/src/facetgroups/CheckboxFacetGroup.js:26]: Thanks for that. I agree with you. Replaced it.

File: packages/scxa-faceted-search-results-v5/src/facetgroups/CheckboxFacetGroupsDefaultProps.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          {
+            group: `isMarkerGenes`,
+            value: `Experiments with marker genes`,
+            label: `Experiments with marker genes`,
+            disabled: false
+          }
+        )
+      }
+
+      return { facets: facets }
```
[the remark this answers is not in the document]
Comment (ke4) [reply packages/scxa-faceted-search-results-v5/src/facetgroups/CheckboxFacetGroupsDefaultProps.js:20]: Thanks for this. I applied it with a bit of modification.

File: packages/scxa-faceted-search-results-v5/package.json [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    "node-fetch": "^2.6.0",
+    "random-words": "1.2.1",
+    "react-test-renderer": "^16.13.1",
+    "webpack": "^4.44.1",
+    "webpack-cli": "^3.3.12",
+    "webpack-dev-server": "^3.11.0"
+  },
+  "scripts": {
+    "test": "echo \"Error: run tests from root\" && exit 1"
+  },
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/package.json:69]: This stub script can go away, or be replaced with something more useful. If you want to use this, and want some inspiration, have a look at [our template package](https://github.com/ebi-gene-expression-group/atlas-package/blob/master/package.json#L11).

File: packages/scxa-faceted-search-results-v5/cypress/component/MultiselectDropdownFacetGroup.cy.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
+import React from "react"
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/cypress/component/MultiselectDropdownFacetGroup.cy.js:1]: Single quotes here.

File: packages/scxa-faceted-search-results-v5/cypress/component/FilterSidebar.cy.js [hunk 11]
```diff
@@ -0,0 +1,260 @@
+import React from "react"
+
+import FilterSidebar from "../../src/FilterSidebar"
+
+import CheckboxFacetGroupsDefaultProps from "../../src/facetgroups/CheckboxFacetGroupsDefaultProps"
+import MultiSelectDropdownFacetGroupsDefaultProps
+  from "../../src/facetgroups/MultiSelectDropdownFacetGroupsDefaultProps"
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/cypress/component/FilterSidebar.cy.js:7]: Single quotes in these imports.

File: packages/scxa-faceted-search-results-v5/cypress/component/FilterList.cy.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import React from "react"
```
Comment (alfonsomunozpomer) [inline packages/scxa-faceted-search-results-v5/cypress/component/FilterList.cy.js:1]: Single quotes.

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
Some style issues and a few opportunities to dry out code. Almost there!

**Review by alfonsomunozpomer — APPROVED** [review]
Good stuff! Two minor corrections and all good.

**Pull request merged**